### DISCUSSION TITLE
How to include page content from other files?

### DISCUSSION THREAD

**Discussion by MiguelVis** [datenstrom/community#691] [Ask a question] [answered]
Hi all.

Is there any way to include a MD file into another one? - ie:

```
This is a MD file.

[include contents.md]

Bye, bye!
```

It could help in multi language sites and making a sort of templates.

**Comment by schulle4u** [comment, marked as answer]
Yes, this is possible. I've published a small [include extension](https://github.com/schulle4u/yellow-include) which I'm using to create overview pages with teasers. You can choose between teaser and full mode. Is this what you are looking for?

**Update**: This extension now has a little less functionality, and only includes content from shared pages folder.

**Reply by MiguelVis** [reply]
Hi @schulle4u.

It looks promising, but I have a problem with relative paths.

It works with absolute paths:

```
[include /es/about/my_name]
```

But it fails with relative paths:

```
[include my_name]
```

Is there any way to solve this?

Thanks!

**Reply by schulle4u** [reply]
Sorry, this is not possible. But actually my extension uses relative URL paths, starting after yur root URL. Can you explain the specific problem you are trying to solve? Maybe the solution is simpler than you think. At least this is my experience with Yellow from the earliest days. 😀

**Reply by MiguelVis** [reply]
What I want is to build a multi language site without the need of replicate **and adjust** every page in each folder.

I though it could be possible to write a common / layout page for each language and copy it in every folder but without have to edit it in order to translate the contents.

I will try to explain it with an example:

```
├── content               
│   ├── 1-en              
│   │   ├── 9-about
│   │   │   ├── page.md    
│   │   │   └── me.md    
│   │   └── shared    
│   ├── 2-de              
│   │   ├── 9-about
│   │   │   ├── page.md    
│   │   │   └── me.md  
│   │   └── shared  
... ...
```

The contents of every `page.md` will be the same, something like:

```
---
Title: About
---
[image my_photo.jpg]

[include me]
```

The contents of each `me.md` should be written in the right language.

Thanks.

**Reply by schulle4u** [reply]
I think you have 2 possibilities for your example. 

1. Use the `page-new*.md` files located in every `shared` folder for each language, e. g. `content/1-en/shared/page-new-default.md`. Whenever you create a new page using your browser, the template content from this file automatically will be inserted. 

2. Using my include extension, create your desired template file in every `shared` folder for each language, e. g. `content/1-en/shared/template.md`. You don't need to enter the full language path, just include it in every page with `[include /shared/template]`, and it should automatically display the right language content. 

HTH 😀

**Reply by markseuffert** [reply]
Additionally to @schulle4u's possibilities:

3. Use `[include /about/me]`. In the multi-language mode the argument will be relative to the current's language home page. Depending on where you are, it will automatically access `/de/about/me`, `/es/about/me`, `/fr/about/me` and so on. Not sure if this is the behaviour is a bug or a feature. 😀

**Reply by MiguelVis** [reply]
> 2. Using my include extension, create your desired template file in every `shared` folder for each language, e. g. `content/1-en/shared/template.md`. You don't need to enter the full language path, just include it in every page with `[include /shared/template]`, and it should automatically display the right language content.

May be I have some settings misconfigured because this does not work in `/es/about/page.md`:

```
[include /shared/me]
```

Results in:

```
ERROR building location '/es/about/', HTTP/1.1 500 Server error: Page '/shared/me' does not exist!
```

But this works perfectly:

```
[include /es/shared/me]
```

In my `yellow-system.ini` I have this:

```
CoreStaticUrl: http://localhost/yellow/public
```

I run Yellow under Xampp and Windows 10.

**Reply by MiguelVis** [reply]
> Additionally to @schulle4u's possibilities:
> 
> 1. Use `[include /about/me]`. In the multi-language mode the argument will be relative to the current's language home page. Depending on where you are, it will automatically access `/de/about/me`, `/es/about/me`, `/fr/about/me` and so on. Not sure if this is the behaviour is a bug or a feature. 😀

Does not work in my case:

```
ERROR building location '/es/about/', HTTP/1.1 500 Server error: Page '/about/me' does not exist!
```

It only works if I prepend `/es`.

I have this in my `yellow-system.ini`:

```
CoreStaticUrl: http://localhost/yellow/public
CoreMultiLanguageMode: 1
```

**Reply by markseuffert** [reply]
You're right. The include extensions needs the exact location.  

If the website home page is at `http://localhost/yellow/public`, use for example:
```
[include /about/me]
[include /es/about/me]
[include /fr/about/me]
```

....or with shared pages:

```
[include /shared/me]
[include /es/shared/me]
[include /fr/shared/me]
```

Tested with Datenstrom Yellow 0.8.18 and Include 0.8.5

**Reply by MiguelVis** [reply]
I'm trying to debug the include extension process and in the function `onParseContentShortcut` in the line:

```
$page = $this->yellow->content->find($location);
```

the parameter `$location` is `/about/me` but later in the function `find` from class `YellowContent` arrives as `/`.

I'm missing something?

**Reply by schulle4u** [reply]
Sorry for the confusion, I've never tested my extensions on multi-language sites to be honest. Currently the include extension doesn't support language detection. I could try to code a workaround to extract language information, but maybe an API change to the `content->find` function would be a more bullet-proof solution?

**Reply by markseuffert** [reply]
Sorry for the confusion previously. Here are some explanations how the [API](https://datenstrom.se/yellow/help/api-for-developers) works. 

Most API functions work with exact locations internally. This means you sometimes build together the location you want before calling API functions. Sometimes you know the exact location and simply call the API functions. There's a lot of flexibility in this design. You can make an extension that normalises relative locations into absolute locations, which is what the Markdown parser does when generating links. You can make an extension that includes child pages only, which is what the blog extension does in the default configuration. You can make an extension that uses pages from a language-specific shared folder, which is how 404 pages are generated in the standard installation.

Here are two examples what the API see in the multi-language mode:

Location = `/yellow/help/api-for-developers`
FileName = `content/1-en/2-yellow/9-help/api-for-developers.md`
URL = `https://datenstrom.se/yellow/help/api-for-developers`

Location = `/de/yellow/help/api-for-developers`
FileName = `content/2-de/2-yellow/9-help/api-for-developers.md`
URL = `https://datenstrom.se/de/yellow/help/api-for-developers`

Not sure if this helps with this question, but hopefully gives an better explanation of what's possible.

**Reply by MiguelVis** [reply]
Hi @schulle4u, @markseu,

Thanks for your help and great effort in order to let me understand how Yellow works.

If I understand correctly, it seems that relative paths do not work well in Yellow when multi-language mode is enabled. Am I right?

That could be a problem sometimes if we don't want reference the absolute path.

**Reply by markseuffert** [reply]
I don't think there's a problem with relative paths or multi-language mode.

Sounds like the include extension doesn't support multi-language websites.

**Reply by markseuffert** [reply]
I think Steffen solved it, the include extension now supports multi-language websites.

**Reply by markseuffert** [reply]
I propose to make this extension available and no longer experimental.

Furthermore, I think the second argument in this shortcut is no longer needed. The reason is that we now have a Markdown syntax for adding CSS to custom block elements. Perhaps it's more readable to add a CSS class or ID in this way. Perhaps it's possible to remove the wrapper `<div class="include>...</div>` as well. After comparing Markdown and AsciiDoc features, one of the use cases that people mentioned was [including files](https://docs.asciidoctor.org/asciidoc/latest/directives/include/). For example to reuse content snippets or code blocks from shared files.

**Reply by schulle4u** [reply]
I agree. I have also removed the div element. This brings the total number of lines down to 23. 🙂 Publication follows, I'll have a last look at the check list to see if nothing missing.
